doc/libi2c.2
.\" man page for Munts Technologies Linux Simple I/O Library
.\"
.\" Copyright (C)2016-2025, Philip Munts dba Munts Technologies.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" * Redistributions of source code must retain the above copyright notice,
.\"   this list of conditions and the following disclaimer.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH libi2c 2 "25 September 2025" "version 1" "Linux Simple I/O Library"
.SH NAME
.B libi2c
\-\- Linux Simple I/O Library: I\u\s-22\s+2\dC Module
.SH SYNOPSIS
.nf
.B #include <libsimpleio/libi2c.h>

.BI "void I2C_open(const char *" name ", int32_t *" fd ", int32_t *" error ");"

.BI "void I2C_close(int32_t " fd ", int32_t *" error ");"

.BI "void I2C_transaction(int32_t " fd ", int32_t " slaveaddr ", void *" cmd ", int32_t " cmdlen ","
.BI "  void *" resp ", int32_t " resplen ", int32_t *" error ");"

.fi
Link with
.BR -lsimpleio .
.SH DESCRIPTION
.nh
All functions return either
.B 0
(upon success) or an
.B errno
value (upon failure) in
.IR *error .
.PP
.B I2C_open()
opens an I\u\s-22\s+2\dC bus controller device.  The device name,
.BR /dev/i2c-x ,
must be passed in the
.IR name
parameter. Upon success, a file descriptor for the
I\u\s-22\s+2\dC bus controller device is returned in
.IR *fd .
.PP
.B I2C_close()
closes a previously opened I\u\s-22\s+2\dC bus controller device.
.PP
.B I2C_transaction()
performs a single I\u\s-22\s+2\dC bus transaction, with optional transmit and receive
phases.  The I\u\s-22\s+2\dC slave device address must be passed in the
.I slaveaddr
parameter.  Either the address of a command message and its length must be passed in the
.IR cmd " and " cmdlen
parameters, or
.BR NULL " and " 0
for a receive only transaction.  Either the address of a receive buffer and its size must be passed in the
.IR resp " and " resplen
parameters, or
.BR NULL " and " 0
for a transmit only transaction.
.SH SEE ALSO
.BR libsimpleio "(2), " libadc "(2), " libdac "(2), " libevent "(2), " libgpio "(2),"
.br
.BR libhidraw "(2), " libiio "(2), " libipv4 "(2)," " liblinux "(2), " liblinx "(2), " libpwm "(2),"
.br
.BR libserial "(2), " libspi "(2), " libstream "(2), " libwatchdog "(2)"
.SH AUTHOR
Philip Munts dba Munts Technologies.


csharp/libI2C.cs
// C# binding for I2C bus controller services in libsimpleio.so

// Copyright (C)2017, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

using System.Runtime.InteropServices;

namespace libsimpleio
{
    /// <summary>
    /// Wrapper for libsimpleio I2C bus controller services.
    /// </summary>
    public class libI2C
    {
        /// <summary>
        /// Open a Linux I2C bus controller device.
        /// </summary>
        /// <param name="devname">Device node name.</param>
        /// <param name="fd">File descriptor.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        [DllImport("simpleio")]
        public static extern void I2C_open(string devname, out int fd,
            out int error);

        /// <summary>
        /// Close a Linux I2C bus controller device.
        /// </summary>
        /// <param name="fd">File descriptor.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        [DllImport("simpleio")]
        public static extern void I2C_close(int fd, out int error);

        /// <summary>
        /// Send bytes to and/or receive bytes from an I2C slave device.
        /// </summary>
        /// <param name="fd">File descriptor.</param>
        /// <param name="slaveaddr">Slave address.</param>
        /// <param name="cmd">Source buffer.</param>
        /// <param name="cmdlen">Source buffer size.</param>
        /// <param name="resp">Response buffer.</param>
        /// <param name="resplen">Response buffer size.</param>
        /// <param name="error">Error code.  Zero upon success or an <code>errno</code>
        /// value upon failure.</param>
        [DllImport("simpleio")]
        public static extern void I2C_transaction(int fd, int slaveaddr,
            byte[] cmd, int cmdlen, byte[] resp, int resplen, out int error);
    }
}
